Libplanet.RocksDBStore/RocksDBStore.cs

```diff
@@ -1361,25 +1361,47 @@ private void LogUnexpectedException(string methodName, Exception e)
             bool includeDeleted
         )
         {
+            long count = 0;
+            Stack<(Guid, long)> chainInfos = new Stack<(Guid, long)>();
+
             if (!includeDeleted && IsDeletionMarked(chainId))
             {
                 yield break;
             }
 
-            long count = 0;
+            chainInfos.Push((chainId, CountIndex(chainId)));
 
-            if (GetPreviousChainInfo(chainId) is { } chainInfo)
+            while (true)
             {
-                Guid prevId = chainInfo.Item1;
-                long pi = chainInfo.Item2;
-
-                int expectedCount = (int)(pi - offset + 1);
-                if (limit is { } limitNotNull && limitNotNull < expectedCount)
+                if (chainInfos.Peek().Item2 > offset &&
+                    GetPreviousChainInfo(chainInfos.Peek().Item1) is { } chainInfo)
                 {
-                    expectedCount = limitNotNull;
+                    chainInfos.Push(chainInfo);
                 }
+                else
+                {
+                    break;
+                }
+            }
+
+            long previousChainTipIndex;
 
-                foreach (BlockHash hash in IterateIndexes(prevId, offset, expectedCount, true))
+            // Adjust offset if it skipped some previous chains.
+            (offset, previousChainTipIndex) =
+                GetPreviousChainInfo(chainInfos.Peek().Item1) is { } cinfo
+                    ? (Math.Max(0, (int)(offset - cinfo.Item2 - 1)), (int)cinfo.Item2)
+                    : (offset, 0);
+
+            while (chainInfos.Count > 0)
+            {
+                (Guid, long) chainInfo = chainInfos.Pop();
+                (Guid cid, long chainTipIndex) = chainInfo;
+
+                // Include genesis block.
+                long expectedCount = chainTipIndex - previousChainTipIndex +
+                                     (GetPreviousChainInfo(cid) is null ? 1 : 0);
```

Comment:
```suggestion
                                     (chainInfos.Count == 0 ? 1 : 0);
```

it seems that we can leverage `chainInfos.Count` instead of `GetPreviousChainInfo()` call.

Reply:
The intention to check `GetPreviousChainInfo(cid) is null` was to know whether the chain part contains the genesis block.

When the `chainInfos.Count` is `0`, it will iterate latest chain part (containing tip index) so it shouldn't add `1` to the count.

For instance, let's assume there are three chain parts, A, B and C. And their tip index is same with the below examples:

- A: 2
- B: 4
- C: 6

When iterating chain `A`, it expects to return `0, 1, 2`.

Or the initial value of `previousChainTipIndex` should be `-1`. 🤔

Reply:
Ah, fair enough. 🆗